lunatrace/bsl/backend/src/github/projects.ts

```diff
@@ -11,42 +11,111 @@
  * limitations under the License.
  *
  */
-import { graphql, GraphqlResponseError } from '@octokit/graphql';
+import { graphql } from '@octokit/graphql';
 import { Configuration, Identity, V0alpha2ApiFactory } from '@ory/kratos-client';
+import deepmerge from 'deepmerge';
 import express from 'express';
 
-import { callHasura } from '../hasura-calls/baseHasuraClient';
-
-const operation = `
-    mutation GetUserAccessToken($user_id: uuid!) {
-        insert_builds_one(object: {project_id: $project_id, s3_url: $s3_url}) {
-          id
-        }
-    }
-`;
-
-export async function getUserGithubAccessToken(variables: { user_id: string }): Promise<string> {
-  const data = await callHasura(operation, 'InsertBuild', variables);
-  if (!data.insert_builds_one) {
-    console.error('Hasura response missing fields ', data);
-    throw new Error('Failed to download manifest for processing');
-  }
-  return '';
-}
+import { hasura } from '../hasura-api';
+import {
+  AuthorizedUserOrganizationsQuery,
+  Identities_Constraint,
+  Identities_Obj_Rel_Insert_Input,
+  Organization_User_Constraint,
+  Organization_User_Insert_Input,
+  Organizations_Constraint,
+  Organizations_Insert_Input,
+  Organizations_Obj_Rel_Insert_Input,
+} from '../hasura-api/generated';
+
+import { GetUserOrganizationsQuery } from './generated';
+import { pullDataForInstallation } from './installation-populate';
+
+import { generateGithubGraphqlClient } from './index';
 
 export const githubApiRouter = express.Router();
 
-export interface KratosIdentityConfig {
-  providers: KratosIdentityProvider[];
-}
+githubApiRouter.get('/webhook/github/install', async (req, res) => {
+  const installationIdQueryParam = req.query.installation_id;
+  const setupActionQueryParam = req.query.setup_action;
+
+  const error = req.query.error;
+  const error_description = req.query.error_description;
+  const error_uri = req.query.error_uri;
+
+  if (error) {
+    res.status(401).send({
+      error: error,
+      description: error_description,
+      uri: error_uri,
+    });
+    return;
+  }
 
-export interface KratosIdentityProvider {
-  initial_id_token: string;
-  subject: string;
-  provider: string;
-  initial_access_token: string;
-  initial_refresh_token: string;
-}
+  if (!installationIdQueryParam || typeof installationIdQueryParam !== 'string') {
+    res.status(401).send({
+      error: 'installation_id not provided in query params',
+    });
+    return;
+  }
+
+  const installationId = parseInt(installationIdQueryParam, 10);
```

Comment:
Nit: You should wrap this in a `try/catch` block or else these exceptions will crash the server (and possibly hang the request)

Reply:
What I usually do in order to reduce nesting is to write a helper function like this:
```ts

interface ParsedInt {
  success: true,
  value: number
};

interface ParsedIntError {
  success: false,
  error: string
}

// By using a union type like this, you're able to avoid possibly "undefined" valued based on if
// success is set to true/false. It's a nice TypeScript feature.
function tryParseInt(value, radix = 10): ParsedInt | ParsedIntError {
  try {
    return {
      value: parseInt(value, radix),
      success: true
    };
  } catch (e) {
    return {
      success: false,
      error: e
    };
  }
}

const {success, value} = tryParseInt('123');

if (!success) {
  res.status(401, {error: 'bad value for installation_id', success: false});
}
```